### PR TITLE
[PyOV] make Tensor responsible for keeping numpy memory

### PR DESCRIPTION
**Pull request by akuporos**
### Details:
 - if `shared_memory=True` and for other ctors which create Tensor with `shared_memory` added ability to keep host memory.
 - if `shared_memory=False` still creates a copy and doesn't tie arrays lifetime to Tensor
 - implemented `custom_keep_alive` (custom calling policy) which adds such ability 

### Tickets:
 - CVS-121493

### CODE REVIEW

File: src/bindings/python/src/pyopenvino/core/tensor.cpp [hunk 1]
```diff
@@ -13,12 +13,23 @@
...
+
 void regclass_Tensor(py::module m) {
-    py::class_<ov::Tensor, std::shared_ptr<ov::Tensor>> cls(m, "Tensor");
+    py::class_<TensorWrapper, std::shared_ptr<TensorWrapper>> cls(m, "Tensor");
     cls.doc() = "openvino.runtime.Tensor holding either copy of memory or shared host memory.";
 
     cls.def(py::init([](py::array& array, bool shared_memory) {
-                return Common::object_from_data<ov::Tensor>(array, shared_memory);
+                auto tensor = Common::object_from_data<ov::Tensor>(array, shared_memory);
+                return TensorWrapper(std::move(tensor), std::move(array));
```
Comment (Wovchena) [inline src/bindings/python/src/pyopenvino/core/tensor.cpp:32]: 1. `TensorWrapper(ov::Tensor&& tensor, py::object array)` doesn't take rvalue to `array`.
2. Given that this lambda takes `py::array& array` by ref, does moving from it affect the actual python object?

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: src/bindings/python/src/pyopenvino/core/tensor.cpp (hunk 1 above)
Comment (akuporos) [reply src/bindings/python/src/pyopenvino/core/tensor.cpp]: it was draft with already outdated code.

File: src/bindings/python/src/pyopenvino/core/tensor.cpp [hunk 2]
```diff
@@ -22,6 +22,7 @@ void regclass_Tensor(py::module m) {
             }),
             py::arg("array"),
             py::arg("shared_memory") = false,
+            py::keep_alive<1, 2>(),
```
Comment (Wovchena) [inline src/bindings/python/src/pyopenvino/core/tensor.cpp:25]: `shared_memory` arg should be removed

Comment (akuporos) [reply src/bindings/python/src/pyopenvino/core/tensor.cpp]: actually we do need `shared_memory` flag, for example for string Tensors -- in such case Tensor may be created only as a copy.
I introduced custom conditional keep_alive which doesn't tie arrays lifetime to Tensor if shared_memory=False

Comment (Wovchena) [reply src/bindings/python/src/pyopenvino/core/tensor.cpp]:
```C++
cls.def(py::init([](py::array& array, const std::vector<size_t> shape, const ov::element::Type& ov_type) {
                return Common::tensor_from_pointer(array, shape, ov_type);
            }))
```
works just like as `shared_memory` was true, but the default value for
```C++
cls.def(py::init([](py::array& array, bool shared_memory) {
                return Common::object_from_data<ov::Tensor>(array, shared_memory);
            }))
```
is `false`.
Switch to `true` to aligning the defaults.

> string Tensors -- in such case Tensor may be created only as a copy.

It's worth describing it in docstrings that `shared_memory=true` raises an error (or ignores the flag) in that case.

**Review by p-wysocki — APPROVED** [review]

File: src/bindings/python/src/pyopenvino/core/tensor.cpp [hunk 3]
```diff
@@ -22,17 +23,14 @@ void regclass_Tensor(py::module m) {
...
                 Tensor's special constructor.
 
                 :param array: Array to create the tensor from.
                 :type array: numpy.array
-                :param shared_memory: If `True`, this Tensor memory is being shared with a host,
-                                      that means the responsibility of keeping host memory is
-                                      on the side of a user. Any action performed on the host
-                                      memory is reflected on this Tensor's memory!
-                                      If `False`, data is being copied to this Tensor.
-                                      Requires data to be C_CONTIGUOUS if `True`.
```
Comment (p-wysocki) [inline src/bindings/python/src/pyopenvino/core/tensor.cpp:35]: Minor nit, shouldn't these two lines stay?

Comment (akuporos) [reply src/bindings/python/src/pyopenvino/core/tensor.cpp:35]: yep, I'll return them back

File: src/bindings/python/src/pyopenvino/core/tensor.cpp (hunk 2 above)
Comment (akuporos) [reply src/bindings/python/src/pyopenvino/core/tensor.cpp]: @Wovchena updated docstring and default value

File: src/bindings/python/src/pyopenvino/core/tensor.cpp (hunk 3 above)
Comment (akuporos) [reply src/bindings/python/src/pyopenvino/core/tensor.cpp]: @p-wysocki returned

**Review by Wovchena — APPROVED** [review]

**Pull request merged**